### PR TITLE
2021-01-31 kickoff

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

Enables us to start developing new API 2020-01-31-preview and test it in dev

### What this PR does / why we need it:

~~This will make an extension to call preview API exclusively.~~
~~If you are using an extension to interact with production RP - this will stop working. Please offload the extension for these activities.~~
Pulled back ^ as we might want to discuss more. But I have a feeling this will be "Told you so". Sometimes we have to break things for sake of progress... 
All E2E in dev will be running on preview API.
All E2E in PROD and INT will be running on stable API

Next will be swagger and cli generation WITH new fields and submission initial swagger for review: https://github.com/Azure/ARO-RP/compare/master...mjudeikis:api.skeleton?expand=1 

### Test plan for issue:

In development

### Is there any documentation that needs to be updated for this PR?

This is just the beginning. Later - yet.

### CODE REVIEW

File: pkg/util/azureclient/mgmt/redhatopenshiftpreview/openshiftclusters.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package redhatopenshiftpreview
```
Comment (jim-minter) [inline pkg/util/azureclient/mgmt/redhatopenshiftpreview/openshiftclusters.go:1]: shall we just admit defeat and go to pkg/util/azureclient/mgmt/redhatopenshift/2021-01-31-preview/redhatopenshift ?

File: pkg/util/cluster/cluster.go [hunk 2]
```diff
@@ -44,15 +46,16 @@ type Cluster struct {
...
-	openshiftclusters redhatopenshift.OpenShiftClustersClient
-	securitygroups    network.SecurityGroupsClient
-	subnets           network.SubnetsClient
-	routetables       network.RouteTablesClient
-	roleassignments   authorization.RoleAssignmentsClient
+	deployments              features.DeploymentsClient
+	groups                   features.ResourceGroupsClient
+	applications             graphrbac.ApplicationsClient
+	serviceprincipals        graphrbac.ServicePrincipalClient
+	openshiftclusters        redhatopenshift.OpenShiftClustersClient
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:53]: call this something awful like openshiftclusters20200430 or something

File: pkg/util/cluster/cluster.go [hunk 3]
```diff
@@ -368,6 +388,54 @@ func (c *Cluster) createCluster(ctx context.Context, clusterName, clientID, clie
 	return c.openshiftclusters.CreateOrUpdateAndWait(ctx, c.ResourceGroup(), clusterName, oc)
 }
 
+func (c *Cluster) createClusterPreview(ctx context.Context, clusterName, clientID, clientSecret string) error {
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:391]: investigate to see if we can define the cluster in the internal representation and use api converters to generate the correct api manifest for the version of client that's being used

File: pkg/util/cluster/cluster.go [hunk 4]
```diff
@@ -270,10 +282,18 @@ func (c *Cluster) Delete(ctx context.Context, clusterName string) error {
...
+		if c.deploymentMode == deployment.Development {
+			err = c.openshiftclusterspreview.DeleteAndWait(ctx, c.ResourceGroup(), clusterName)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		} else {
+			err := c.openshiftclusters.DeleteAndWait(ctx, c.ResourceGroup(), clusterName)
+			if err != nil {
+				errs = append(errs, err)
+			}
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:294]: I'm not sure this is worth it

File: python/az/aro/azext_aro/_client_factory.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 import urllib3
 
 from azext_aro.custom import rp_mode_development
-from azext_aro.vendored_sdks.azure.mgmt.redhatopenshift.v2020_04_30 import AzureRedHatOpenShiftClient
+from azext_aro.vendored_sdks.azure.mgmt.redhatopenshift.v2021_01_31_preview import AzureRedHatOpenShiftClient
```
Comment (jim-minter) [inline python/az/aro/azext_aro/_client_factory.py:7]: Let's think about and discuss the az side of this a bit more.

File: pkg/util/cluster/cluster.go [hunk 6]
```diff
@@ -22,7 +22,11 @@ import (
 	uuid "github.com/satori/go.uuid"
 	"github.com/sirupsen/logrus"
 
-	mgmtredhatopenshift "github.com/Azure/ARO-RP/pkg/client/services/redhatopenshift/mgmt/2020-04-30/redhatopenshift"
+	"github.com/Azure/ARO-RP/pkg/api"
+	v20200430 "github.com/Azure/ARO-RP/pkg/api/v20200430"
+	v2021131preview "github.com/Azure/ARO-RP/pkg/api/v20210131preview"
+	mgmtopenshiftclustersv20200430 "github.com/Azure/ARO-RP/pkg/client/services/redhatopenshift/mgmt/2020-04-30/redhatopenshift"
+	mgmtopenshiftclustersv20210131preview "github.com/Azure/ARO-RP/pkg/client/services/redhatopenshift/mgmt/2021-01-31-preview/redhatopenshift"
```
Comment (mjudeikis) [inline pkg/util/cluster/cluster.go:29]: I dont know what is better. Keeping this or having createCluster 2 methods...

File: pkg/util/cluster/cluster.go [hunk 7]
```diff
@@ -317,52 +324,82 @@ func (c *Cluster) Delete(ctx context.Context, clusterName string) error {
...
+			return err
+		}
+
+		ocExt := mgmtopenshiftclustersv20200430.OpenShiftCluster{}
+		err = json.Unmarshal(data, &ocExt)
+		if err != nil {
+			return err
+		}
+
+		return c.openshiftclustersv20200430.CreateOrUpdateAndWait(ctx, c.env.ResourceGroup(), clusterName, ocExt)
```
Comment (mjudeikis) [inline pkg/util/cluster/cluster.go:401]: This is where create magic happens.

File: python/az/aro/azext_aro/custom.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import random
 import os
 
-import azext_aro.vendored_sdks.azure.mgmt.redhatopenshift.v2020_04_30.models as v2020_04_30
+import azext_aro.vendored_sdks.azure.mgmt.redhatopenshift.v2020_04_30.models as openshiftcluster
```
Comment (mjudeikis) [inline python/az/aro/azext_aro/custom.py:7]: Will make future renames easier when we agree what we want to do with this

File: test/e2e/adminapi_resources.go [hunk 9]
```diff
@@ -26,7 +26,7 @@ var _ = Describe("[Admin API] List Azure resources action", func() {
 		resourceID := resourceIDFromEnv()
 
 		By("getting the resource group where cluster resources live in")
-		oc, err := clients.OpenshiftClusters.Get(ctx, _env.ResourceGroup(), clusterName)
+		oc, err := clients.OpenshiftClustersv20200430.Get(ctx, _env.ResourceGroup(), clusterName)
```
Comment (mjudeikis) [inline test/e2e/adminapi_resources.go:29]: For now E2E are using old API. So this basically scatters our e2e. Cluster will be created with new API. So any new future tests should be not API binded. In example.
If we create FIPs cluster, we should write tests which currently run in dev only and uses Kubernetes API to check it. Not ARO... 

:hankey: a bit. Not sure how we can do this without copy-paste test cases.

File: swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2021-01-31-preview/examples/OpenShiftClusters_CreateOrUpdate.json [hunk 10]
```diff
@@ -0,0 +1,162 @@
+{
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2021-01-31-preview/examples/OpenShiftClusters_CreateOrUpdate.json:1]: I think these files should be going into swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview

File: swagger/redhatopenshift/resource-manager/readme.md [hunk 11]
```diff
@@ -38,6 +38,15 @@ input-file:
   - Microsoft.RedHatOpenShift/stable/2020-04-30/redhatopenshift.json
 ```
 
+### Tag: package-2021-01-31
+
+These settings apply only when `--tag=package-2021-01-31` is specified on the command line.
+
+``` yaml $(tag) == 'package-2021-01-31'
+input-file:
+  - Microsoft.RedHatOpenShift/stable/2021-01-31/redhatopenshift.json
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/readme.md:47]:
```suggestion
  - Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json
```

File: swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2021-01-31-preview/redhatopenshift.json [hunk 12]
```diff
@@ -0,0 +1,796 @@
+{
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/stable/2021-01-31-preview/redhatopenshift.json:1]: should be swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/redhatopenshift.json

File: swagger/redhatopenshift/resource-manager/readme.go.md [hunk 13]
```diff
@@ -24,3 +25,12 @@ Please also specify `--go-sdk-folder=<path to the root directory of your azure-s
 ``` yaml $(tag) == 'package-2020-04-30' && $(go)
 output-folder: $(go-sdk-folder)/services/$(namespace)/mgmt/2020-04-30/$(namespace)
 ```
+
+### Tag: package-2021-01-31 and go
+
+These settings apply only when `--tag=package-2021-01-31 --go` is specified on the command line.
```
Comment (jim-minter) [inline swagger/redhatopenshift/resource-manager/readme.go.md:31]:
```suggestion
These settings apply only when `--tag=package-2021-01-31-preview --go` is specified on the command line.
```
etc (multiple changes)

File: test/e2e/adminapi_resources.go (hunk 9 above)
Comment (jim-minter) [reply test/e2e/adminapi_resources.go:29]: I don't agree with this or see the problem.  Let's discuss tomorrow.

**Review by jim-minter — CHANGES_REQUESTED** [review]
Minor fixes requested. I'm OK with this.

**Review by jim-minter — APPROVED** [review]
I believe this is good, if @mjudeikis does.

**Pull request merged**